wi-fi/hal/cyhal_sdio.c

```diff
@@ -0,0 +1,783 @@
+			(cyhal_utils_set_iomux(pctl_mux_csi_d3, 1) < 0)) {    /* USDHC2_DATA3 */
+			cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "can't configure SDIO pins iomux\n");
+			break;
+		}
+
+		if ((cyhal_utils_set_iosel(pctl_isel_usdhc2_clk, 0) < 0) || /* CSI_VSYNC_ALT1 */
+			(cyhal_utils_set_iosel(pctl_isel_usdhc2_cmd, 0) < 0) || /* CSI_HSYNC_ALT1 */
+			(cyhal_utils_set_iosel(pctl_isel_usdhc2_d0, 0) < 0) ||  /* CSI_DATA00_ALT1 */
+			(cyhal_utils_set_iosel(pctl_isel_usdhc2_d1, 0) < 0) ||  /* CSI_DATA01_ALT1 */
+			(cyhal_utils_set_iosel(pctl_isel_usdhc2_d2, 2) < 0) ||  /* CSI_DATA02_ALT1 */
+			(cyhal_utils_set_iosel(pctl_isel_usdhc2_d3, 0) < 0)) {  /* CSI_DATA03_ALT1 */
+			cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "can't configure SDIO pins iosel\n");
+			break;
+		}
+
+		if ((cyhal_utils_set_iopad(pctl_pad_csi_vsync, 0, 0, 0, 0, 0, 2, 1, 0) < 0) ||
+			(cyhal_utils_set_iopad(pctl_pad_csi_hsync, 0, 2, 1, 1, 0, 2, 1, 0) < 0) || /* 100K Ohm Pull up */
+			(cyhal_utils_set_iopad(pctl_pad_csi_d0, 0, 2, 1, 1, 0, 2, 1, 0) < 0) ||    /* 100K Ohm Pull up */
+			(cyhal_utils_set_iopad(pctl_pad_csi_d1, 0, 2, 1, 1, 0, 2, 1, 0) < 0) ||    /* 100K Ohm Pull up */
+			(cyhal_utils_set_iopad(pctl_pad_csi_d2, 0, 2, 1, 1, 0, 2, 1, 0) < 0) ||    /* 100K Ohm Pull up */
+			(cyhal_utils_set_iopad(pctl_pad_csi_d3, 0, 2, 1, 1, 0, 2, 1, 0) < 0)) {    /* 100K Ohm Pull up */
+			cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "can't configure SDIO pins iopad\n");
+			break;
+		}
+
+		/* enable USDHC2_CLK */
+		if (cyhal_utils_set_devclk(pctl_clk_usdhc2, 3) < 0) {
+			cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "can't enable SDIO clock\n");
+			break;
+		}
+
+		/* reset SDIO and set SDIO clock freq to 400000000 / 2 / (256 * 2) = 390625 Hz */
+		reset_all(0x80, 0x1); /* SDCLKFS=0x80 DVS=0x1 */
+
+		/* change card detection pin polarity */
+		val = *(sdio_common.base + vend_spec);
+		val |= (1 << 5); /* CD_POL=1 */
+		*(sdio_common.base + vend_spec) = val;
+
+		/* DMAESEN=1 DEBESEN=1 DCESEN=1 DTOESEN=1 CIESEN=1 CEBESEN=1 CCESEN=1 CTOESEN=1 DINTSEN=1 BGESEN=1 TCSEN=1 CCSEN=1 */
+		*(sdio_common.base + int_status_en) = 0x107f000f;
+
+		// dump_registers();
+
+		return CY_RSLT_SUCCESS;
+	} while (0);
+
+	cyhal_sdio_free(obj);
+
+	return CYHAL_SDIO_RSLT_ERR_BAD_PARAM;
+}
+
+
+/* NOTE: obj is ignored - state is kept in sdio_common */
+void cyhal_sdio_free(cyhal_sdio_t *obj)
+{
+	cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "cyhal_sdio_free\n");
+
+	/* reset SDIO and set SDIO clock freq to 400000000 / 2 / (256 * 2) = 390625 Hz */
+	reset_all(0x80, 0x1); /* SDCLKFS=0x80 DVS=0x1 */
+
+	if (sdio_common.irq_thread_id) {
+		/* request IRQ thread finish */
+		mutexLock(sdio_common.irq_lock);
+		sdio_common.irq_thread_finish = true;
+		condSignal(sdio_common.irq_cond);
+		mutexUnlock(sdio_common.irq_lock);
+
+		/* wait for IRQ thread finish */
+		mutexLock(sdio_common.irq_lock);
+		while (sdio_common.irq_thread_finish) {
+			condWait(sdio_common.irq_cond, sdio_common.irq_lock, 0);
+		}
+		mutexUnlock(sdio_common.irq_lock);
```

Comment:
missing thread join - memory leak

Reply:
Yep, I thought it was not implemented in RTOS.